### ISSUE TITLE
How to ignore device files when searching all file system for something?

### ISSUE THREAD

**Issue by beppe9000** [ggreer/the_silver_searcher#1339]
I would like ag to automatically avoid reading these kind of files...

`ERR: File ./sys/devices/pci0000:00/0000:00:02.0/graphics/fb0/uevent failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/graphics/fb0/virtual_size failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/power/control failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/power/async failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/power/runtime_enabled failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/power/runtime_active_kids failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/power/runtime_active_time failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/power/autosuspend_delay_ms failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/power/runtime_status failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/power/runtime_usage failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/power/runtime_suspended_time failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/resource failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/driver_override failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/consistent_dma_mask_bits failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/modalias failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/dma_mask_bits failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/graphics/fb0/stride failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/local_cpus failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/config failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/device failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/enable failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/local_cpulist failed to load: No such device.
ERR: Skipping ./sys/devices/pci0000:00/0000:00:02.0/remove: Error opening file: Permission denied
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/msi_bus failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/uevent failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/vendor failed to load: No such device.
ERR: Skipping ./sys/devices/pci0000:00/0000:00:02.0/rescan: Error opening file: Permission denied
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/subsystem_device failed to load: No such device.
ERR: File ./sys/devices/pci0000:00/0000:00:02.0/boot_vga failed to load: No such device.`

**Comment by genivia-inc** [comment]
Just to chime in, since this issue is a valid one that can be easily avoided. I came up with an efficient approach to restrict searches to certain file systems, see [Including or excluding mounted file systems from searches](https://github.com/Genivia/ugrep#fs):
~~~
--exclude-fs=MOUNTS
        Exclude file systems specified by MOUNTS from recursive searches,
        MOUNTS is a comma-separated list of mount points or pathnames of
        directories on file systems.  Note that --exclude-fs mounts take
        priority over --include-fs mounts.  This option may be repeated.
--include-fs=MOUNTS
        Only file systems specified by MOUNTS are included in recursive
        searches.  MOUNTS is a comma-separated list of mount points or
        pathnames of directories on file systems.  --include-fs=. restricts
        recursive searches to the file system of the working directory
        only.  Note that --exclude-fs mounts take priority over
        --include-fs mounts.  This option may be repeated.
~~~
Basically, option `--include-fs=.` restricts the search to the file system associated with the working directory. But you can specify any directory or the mount point(s) directly.

The implementation is fairly straightforward actually. Just use `statvfs` to obtain the FS identifier of the given dir as an argument to `--include-fs`, and compare it to the dirs and devices before descending into those recursively, when option `-r` or `-R` is used.

**Comment by AnrDaemon** [comment]
The generic "stay on same filesystem" flag is much more useful.
